enhancements/storage/csi-migration.md

```diff
@@ -0,0 +1,246 @@
+
+This patch would be carried over in OCP until all in-tree storage plugins are migrated to CSI, which should happen around 3 or 4 releases.
+Initially we would start with Cinder and GCP (_CSIMigrationGCE_ and _CSIMigrationCinder_ flags), so that we are aligned with the goals of both upstream and [CCMO](https://github.com/openshift/enhancements/pull/463).
+In subsequent OCP releases we would proceed with EBS, vSphere and Azure (_CSIMigrationAWS_, _CSIMigrationvSphere_ and _CSIMigrationAzureFile_), without a predefined order.
+
+With this patch, when deciding about using either the CSI driver or the in-tree plugin, the Attach Detach Controller will **only** rely on the information propagated by the node. In other words, Attach Detach Controller will start considering which plugin to use (in-tree or CSI) on a node basis only.
+
+Note: one might wonder why upstream does not implement this behaviour already. The answer is that they do not have the same issues OCP has in regards to disabling feature flags in the wrong order across components.
+
+#### Benefits
+
+The biggest benefit of this approach is that we do not need to worry about the ordering in which components are restarted with the CSI migration flags switched on or off. Migration flags can be switched on or off across components in any order.
+
+Another benefit is that this approach should not break downgrades once CSI Migration becomes GA. In OCP, a downgrade is fundamentally a regular upgrade to an older version.
+That means that CVO downgrades components in the same order as it upgrades them: control-plane first, nodes later. This would impose an issue if the user downgraded from a version with CSI Migration enabled to a version with CSI Migration disabled. With the above patch in the downgraded version, that would not be a problem.
+
+It is important to notice that, with this carry-patch in OCP, Attach Detach Controller will _not_ change its current behaviour as long as nodes are not migrated to CSI, which is the default behaviour in OCP 4.8.
+
+#### Graduation
+
+During the Tech Preview phase in OCP 4.8, users can enable and disable the CSI migration feature using the **TechPreviewNoUpgrade** *FeatureSet*. Among other things, this *FeatureSet* switches the complete [Cloud Migration](https://github.com/openshift/enhancements/pull/463) in OCP 4.8. Here is a configuration example:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: FeatureGate
+metadata:
+  name: cluster
+spec:
+  featureSet: TechPreviewNoUpgrade
+```
+
+Optionally, users can enable the CSI Migration *without* enabling the Cloud Migration by using the **CustomNoUpgrade** *FeatureSet*. Here is a configuration example to enable the migration for AWS EBS:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: FeatureGate
+metadata:
+  name: cluster
+spec:
+  featureSet: CustomNoUpgrade
+  customNoUpgrade:
+    enabled:
+    - CSIMigrationAWS
+```
```

Comment:
@bertinatto Hi, I have a question about the `CustomNoUpgrade` and `TechPreviewNoUpgrade` featureset. From the api file I can see:
```
        // TechPreviewNoUpgrade turns on tech preview features that are not part of the normal supported platform. Turning
        // this feature set on CANNOT BE UNDONE and PREVENTS UPGRADES.
        TechPreviewNoUpgrade FeatureSet = "TechPreviewNoUpgrade"

        // CustomNoUpgrade allows the enabling or disabling of any feature. Turning this feature set on IS NOT SUPPORTED, CANNOT BE UNDONE, and PREVENTS UPGRADES.
        // Because of its nature, this setting cannot be validated.  If you have any typos or accidentally apply invalid combinations
        // your cluster may fail in an unrecoverable way.
        CustomNoUpgrade FeatureSet = "CustomNoUpgrade"
```
Looks like `CustomNoUpgrade` is not supported? Both featuresets can be `UNDONE` and prevent `UPGRADE`, does this mean the featuresets can not be disabled and will block the cluster upgrade?

Reply:
Yes, CustomNoUpgrade is not supported. We support TechPreviewNoUpgrade, but only as tech preview. However, feedback from any of them is always welcome - they may show a bug that needs to be fixed before GA.

And yes, cannot be undone and prevents upgrade is correct. In case of CSI migration I think it could be undone, but every tech preview feature behind this FeatureSet is different and it's safer to say we officially don't allow to go back to the fully supported state.

Reply:
I think we plan to support upgrade and undone for the csi migration. So maybe the original design to create a new featureset is better?

Reply:
@qinpingli we do want to support upgrades and undo, but only when the CSI Migration for the given plugin is GA.

However, for Tech Preview it's OK to no support upgrades and we expect users to test the feature in throw-away clusters. In addition to that, we need to enable the CSI Migration once the users enables the [Cloud Migration](https://github.com/openshift/enhancements/pull/463), and CCMO uses the `TechPreviewNoUpgrade` FeatureSet.

In theory, we _could_ create a separate FeatureSet for CSI Migration that supports upgrades and undo during Tech Preview. We'll discuss that with the team.